test/src/test/groovy/org/openremote/test/protocol/http/HttpClientProtocolTest.groovy

```diff
@@ -549,4 +563,136 @@ class HttpClientProtocolTest extends Specification implements ManagerContainerTr
             HTTPProtocol.client.set(null)
         }
     }
+
+    def "Check HTTP client dynamic time feature"() {
+
+        given: "expected conditions"
+        def conditions = new PollingConditions(timeout: 10, initialDelay: 1)
+
+        and: "the HTTP client protocol min times are adjusted for testing"
+        HTTPProtocol.MIN_POLLING_MILLIS = 10
+
+        and: "the container starts"
+        def container = startContainer(defaultConfig(), defaultServices())
+        def assetStorageService = container.getService(AssetStorageService.class)
+        def assetProcessingService = container.getService(AssetProcessingService.class)
+        def agentService = container.getService(AgentService.class)
+        def timerService = container.getService(TimerService.class)
+
+//        timerService.stop();
+
+        when: "the web target builder is configured to use the mock server"
+        HTTPProtocol.initClient()
+        if (!HTTPProtocol.client.get().configuration.isRegistered(mockServer)) {
+            HTTPProtocol.client.get().register(mockServer, Integer.MAX_VALUE)
+        }
+
+        and: "a HTTP client agent is created"
+
+        String testReplacementString = "\${time:PT1H:yyyy-MM-dd HH:mm}"
+        HTTPAgent agent = new HTTPAgent("Test agent")
+                .setRealm(Constants.MASTER_REALM)
+                .setBaseURI("https://mockapi")
+                .setOAuthGrant(
+                        new OAuthPasswordGrant("https://mockapi/token",
+                                "TestClient",
+                                "TestSecret",
+                                "scope1 scope2",
+                                "testuser",
+                                "password")
+                )
+                .setFollowRedirects(true)
+
+        and: "the agent is added to the asset service"
+        agent = assetStorageService.merge(agent)
+
+        then: "the protocol should authenticate and the connection status should become CONNECTED"
+        conditions.eventually {
+            agent = assetStorageService.find(agent.id, HTTPAgent.class)
+            assert agent.getAgentStatus().orElse(ConnectionStatus.DISCONNECTED) == ConnectionStatus.CONNECTED
+        }
+
+        when: "An array of dynamic time strings is tested with the string replacement function"
+
+        String[] dynamicTimeTests = ["\${time:PT1H:yyyy-MM-dd HH:mm}", "\${time:PT2H:yyyy-MM-dd HH:mm:ss}", "\${time:P1D:yyyy-MM-dd}", "\${time:P2DT3H:yyyy-MM-dd HH:mm}", "\${time:PT30M:yyyy-MM-dd HH:mm}", "\${time:P7D:yyyy-MM-dd}", "\${time:PT5H:yyyy-MM-dd HH:mm}", "\${time:P2D:yyyy-MM-dd}", "\${time:PT1M:yyyy-MM-dd}", "\${time:P3D:yyyy-MM-dd}"] as List<String>
+
+        then: "The strings are parsed correctly. If there is a way to use a constant timestamp, it'd be much easier."
+        for (final String e in dynamicTimeTests){
```

Comment:
Not a big fan of unit testing like this esp inside an integration test. As this is doing nothing then it should be removed.

Reply:
Agreed. The comment about having the timer be stable was so that I can generate a list of test cases and their predefined 100% correct results, and then comparing each other. But I think that an integration test that covers the required test cases should be enough.